lib/sinatra/base.rb

```diff
@@ -1016,10 +1016,8 @@ def route_missing
     # a matching file is found, returns nil otherwise.
     def static!
       return if (public_dir = settings.public_folder).nil?
-      public_dir = File.expand_path(public_dir)
-
-      path = File.expand_path(public_dir + unescape(request.path_info))
-      return unless path.start_with?(public_dir) and File.file?(path)
```

Comment:
In case anyone else looks at this diff and says "this looks like it adds a directory traversal vulnerability", it doesn't. The default `rack-protection` middleware mitigates directory traversal [here](https://github.com/sinatra/rack-protection/blob/master/lib/rack/protection/path_traversal.rb).